transformer_contextual_lenses.py

```diff
@@ -0,0 +1,706 @@
+# -*- coding: utf-8 -*-
+"""transformer_contextual_lenses.ipynb
+
+Automatically generated by Colaboratory.
+
+Original file is located at
+    https://colab.research.google.com/drive/16NcwXKxsWDshonpEciQox8I3dSDgEH9w
+
+# Initialization
+"""
+
+# *** Comment out on cloud ***
+# Install the newest JAX and FLAX versions.
+!pip install --upgrade -q jax==0.1.61 jaxlib==0.1.42 flax==0.1.0rc2
+
+# *** Comment out on cloud ***
+import requests
+import os
+if 'TPU_DRIVER_MODE' not in globals():
+  url = 'http://' + os.environ['COLAB_TPU_ADDR'].split(':')[0] + ':8475/requestversion/tpu_driver_nightly'
+  resp = requests.post(url)
+  TPU_DRIVER_MODE = 1
+
+# The following is required to use TPU Driver as JAX's backend.
+from jax.config import config
+config.FLAGS.jax_xla_backend = "tpu_driver"
+config.FLAGS.jax_backend_target = "grpc://" + os.environ['COLAB_TPU_ADDR']
+print(config.FLAGS.jax_backend_target)
+
+import functools
+import itertools
+import os
+import time
+
+import flax
+from flax import jax_utils
+from flax import nn
+from flax import optim
+from flax.metrics import tensorboard
+from flax.training import checkpoints
+from flax.training import common_utils
+
+import jax
+from jax import random
+from jax import lax
+import jax.nn
+import jax.numpy as jnp
+
+import numpy as np
+
+import matplotlib.pyplot as plt
+
+"""# Transformer model
+Code source: https://github.com/google/flax/blob/master/examples/lm1b/models.py
+"""
+
+def shift_right(x):
+  """Shift the input to the right by padding on axis 1."""
+  pad_widths = [(0, 0)] * len(x.shape)
+  pad_widths[1] = (1, 0)  # Padding on axis=1
+  padded = jnp.pad(
+      x, pad_widths, mode='constant', constant_values=x.dtype.type(0))
```

Comment:
Not a bug in current usage, but something to be aware of: Both PAD and BOS (beginning of sentence) are 0 in the LM1B example (and therefore this fork). This is prone to bugs down the line.